scl/paloalto/panos.conf

```diff
@@ -0,0 +1,273 @@
+          template("${`prefix`tmp}")
+          delimiters(',')
+        );
+      };
+      filter { "${`prefix`device_name}" ne "" };
+    } 
+    elif (match('CORRELATION' value('`prefix`type') type(string))) {
+      parser {
+
+        # FUTURE_USE, Receive Time, Serial Number, Type, Content/Threat
+        # Type, FUTURE_USE, Generated Time, Source Address.  Source User,
+        # Virtual System, Category, Severity, Device Group Hierarchy Level
+        # 1, Device Group Hierarchy Level 2, Device Group Hierarchy Level 3,
+        # Device Group Hierarchy Level 4, Virtual System Name, Device Name,
+        # Virtual System ID, Object Name, Object ID, Evidence
+
+        csv-parser(
+          columns("src","srcuser","vsys","category","severity",
+                  "dg_hier_level_1","dg_hier_level_2","dg_hier_level_3","dg_hier_level_4","vsys_name","device_name",
+                  "vsys_id","objectname","object_id","evidence")
+          prefix("`prefix`")
+          template("${`prefix`tmp}")
+          delimiters(',')
+        );
+      };
+      filter { "${`prefix`device_name}" ne "" };
+    } 
+    elif (match('USERID' value('`prefix`type') type(string))) {
+      parser {
+        csv-parser(
+
+          # FUTURE_USE, Receive Time, Serial Number, Type, Threat/Content
+          # Type, FUTURE_USE, Generated Time, Virtual System, Source IP,
+          # User, Data Source Name, Event ID, Repeat Count, Time Out
+          # Threshold, Source Port, Destination Port, Data Source, Data
+          # Source Type, Sequence Number, Action Flags, Device Group
+          # Hierarchy Level 1, Device Group Hierarchy Level 2, Device Group
+          # Hierarchy Level 3, Device Group Hierarchy Level 4, Virtual
+          # System Name, Device Name, Virtual System ID, Factor Type, Factor
+          # Completion Time, Factor Number, FUTURE_USE, FUTURE_USE, User
+          # Group Flags, User by Source
+
+          columns("vsys","ip","user","datasourcename","eventid","repeatcnt","timeout","beginport","endport","datasource","datasourcetype","seqno","actionflags",
+                  "dg_hier_level_1","dg_hier_level_2","dg_hier_level_3","dg_hier_level_4","vsys_name","device_name",
+                  "vsys_id","factortype","factorcompletiontime","factorno","future_use3","future_use4","ugflags","userbysource")
+          prefix("`prefix`")
+          template("${`prefix`tmp}")
+          delimiters(',')
+        );
+      };
+      filter { "${`prefix`device_name}" ne "" };
+    };
+
+    # no else branch, so we would drop messages where type does not match
+    # any of the known ones in the branches above
+
+    rewrite {
+
+      # fix the message in case we forward it to a syslog consumer
+      set("${LEGACY_MSGHDR}${MESSAGE}" value("MESSAGE"));
+      set("paloalto_panos" value("PROGRAM"));
+      unset(value("`prefix`tmp"));
+    };
+
+  };
+};
+application panos[syslog] {
+    filter { match("1," value("PROGRAM") type(string) flags(prefix)); };
```

Comment:
According to https://docs.paloaltonetworks.com/pan-os/8-1/pan-os-admin/monitoring/use-syslog-for-monitoring/syslog-field-descriptions/config-log-fields,
the first field is
```
Serial Number (serial) | Serial number of the device that generated the log.
```
and this is where the `1,` comes from? Originally we checked for the paloalto string in the message. How are these serial numbers are assigned to the devices? Are we safe to assume all devices will have `1` as serial number? Would it be better check if `PROGRAM` contains a `,`: like `.*,` instead? That would be more permissive, and with `drop-invalid` flags in the csv parser, we would still be safe-ish against false positives.

Reply:
Yes thank you! I need new glasses... It is the `FUTURE_USE`. All sounds great then.